### PR TITLE
feat(esptool): allow filtering ports by VID/PID (ESPTOOL-878)

### PR DESCRIPTION
**Pull request by bryghtlabs-richard**
Enables user to specify --usbVid and/or --usbPid, which filter the com port list before checking for ESP chips. With many serial ports on the computer, this greatly speeds up the process of esptool identifying the correct com port by allowing esptool to skip all com ports without a matching VID/PID. Implements https://github.com/espressif/esptool/issues/987

# I have tested this change with the following hardware & software combinations:

NO TESTING ON MAC/LINUX. Tested on Windows64 with ESP32-S3-DevKitC and custom ESP32-S3 board.

```
$ time esptool read_flash_status
esptool.py v4.7.0
Found 4 serial ports
Serial port COM30
Connecting......................................
COM30 failed to connect: Failed to connect to Espressif device: No serial data received.
For troubleshooting steps visit: https://docs.espressif.com/projects/esptool/en/latest/troubleshooting.html
Serial port COM29
Connecting......................................
COM29 failed to connect: Failed to connect to Espressif device: No serial data received.
For troubleshooting steps visit: https://docs.espressif.com/projects/esptool/en/latest/troubleshooting.html
Serial port COM18
Connecting...
Detecting chip type... ESP32-S3
...long text omitted...
real    0m18.193s
user    0m0.000s
sys     0m0.015s
```

```
$ time esptool --usbVid 0x303a read_flash_status
esptool.py v4.7.0
Found 1 serial ports
Serial port COM18
Connecting...
Detecting chip type... ESP32-S3
...long text omitted...
real    0m1.067s
user    0m0.000s
sys     0m0.000s
```

# I have run the esptool.py automated integration tests with this change and the above hardware:

$ pytest test_esptool.py --port COM49 --chip esp32s3 --baud 230400
============================= test session starts =============================
platform win32 -- Python 3.11.5, pytest-8.2.2, pluggy-1.5.0
rootdir: C:\Projects\esptool
configfile: pyproject.toml
collected 88 items

test_esptool.py sss.......ss.................s.s.........s....sss.s..... [ 63%]
......s...sssssss..ssss..s..ssss                                         [100%]

============================== warnings summary ===============================
test_esptool.py:120
  C:\Projects\esptool\test\test_esptool.py:120: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=1, condition=arg_preload_port is not False)

test_esptool.py:1204
  C:\Projects\esptool\test\test_esptool.py:1204: PytestUnknownMarkWarning: Unknown pytest.mark.flaky - is this a typo?  You can register custom marks to avoid this warning - for details, see https://docs.pytest.org/en/stable/how-to/mark.html
    @pytest.mark.flaky(reruns=5)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========== 59 passed, 29 skipped, 2 warnings in 415.91s (0:06:55) ============

### CODE REVIEW

**Comment by dobairoland** [comment]
Thank you for the contribution. I think this is a good idea but we might think about a little what new CLI arguments we add. I'm worried that others might need other type of filtering (e.g. serial_number, location) for which we would have to add also new arguments. 

- This is one solution to merge the arguments: `--port-filter <vid>:<pid>` where `<vid>` and `<pid>` could be `*` as well. 
- An even more general solution: `--port-filter vid=<number> --port-filter pid=<number>` where `--port-filter` can be repeated several times and can be easily extended.

What do you think @radimkarnis, @peterdragun?

**Comment by peterdragun** [comment]
Overall I like the idea, but as mentioned I would go with one argument for both options.
Maybe we can extend the filter to also cover the port name? For example on macOS ports can have different names based on whether they are using a USB to UART bridge (`/dev/cu.usbserial-XX`) or USB Serial/JTAG (`/dev/cu.usbmodemXX`). So in case I know which peripheral I want to use but don't know the number it may be helpful. I mean we can do that based on VID as well, but it would be easier to remember the port name than VID.

**Comment by dobairoland** [comment]
> Overall I like the idea, but as mentioned I would go with one argument for both options. Maybe we can extend the filter to also cover the port name? For example on macOS ports can have different names based on whether they are using a USB to UART bridge (`/dev/cu.usbserial-XX`) or USB Serial/JTAG (`/dev/cu.usbmodemXX`). So in case I know which peripheral I want to use but don't know the number it may be helpful. I mean we can do that based on VID as well, but it would be easier to remember the port name than VID.

I think this could be covered by the following if we go with my second proposal:
```
--port-filter name=usbserial
```

**Comment by bryghtlabs-richard** [comment]
> An even more general solution: --port-filter vid=<number> --port-filter pid=<number> where --port-filter can be repeated several times and can be easily extended.

I like this.

> I think this could be covered by the following if we go with my second proposal: `--port-filter name=usbserial`

Are you thinking a substring search, or should we check against a prefix from the pyserial port name, like `name=/dev/cu.usbserial`?

**Comment by dobairoland** [comment]
> Are you thinking a substring search, or should we check against a prefix from the pyserial port name, like name=/dev/cu.usbserial?

Substring search is what I had in my mind. `ttyUSB` and `ttyACM` could be easily used on Linux, or even simply `USB` or `ACM`. But this was just a suggestion so we don't end up needing several similar arguments. You don't have to address this in this PR (if you don't want).

On the other hand, what would be great to update in this PR is the documentation and mention this new feature somewhere. Perhaps in a separate section in https://docs.espressif.com/projects/esptool/en/latest/esp32/esptool/advanced-options.html.

**Comment by dobairoland** [comment]
Hi @bryghtlabs-richard. Could you please get back to this pull request and finish it as we've discussed?

**Comment by bryghtlabs-richard** [comment]
Hi @dobairoland , I'm back from a vacation and I think I've got everything we want, and updated the documentation.

File: docs/en/esptool/advanced-options.rst [hunk 1]
```diff
@@ -117,3 +117,18 @@ An example of this is available in the :ref:`merge_bin <merge-bin>` command desc
 .. note:: PowerShell users
 
     Because of `splatting <https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_splatting?view=powershell-7.3>`__ in PowerShell (method of passing a collection of parameter values to a command as a unit) there is a need to add quotes around @filename.txt ("@filename.txt") to be correctly resolved.
+
+Filtering serial ports
+----------------------
+.. _filtering_serial_ports:
+
+``--port-filter <FilterType>=<FilterValue>`` allows limiting ports that will be tried. This can be useful when esptool is run on a system
+with many serial ports There are a few different types that can be combined. A port must match all specified FilterTypes, and must match
```
Comment (dobairoland) [inline docs/en/esptool/advanced-options.rst:126]: "ports There" -> "ports. There"

File: esptool/__init__.py [hunk 2]
```diff
@@ -1010,21 +1035,34 @@ def arg_auto_chunk_size(string: str) -> int:
...
+    for port in list_ports.comports():
+        if sys.platform == "darwin" and port.device.endswith(
+            ("Bluetooth-Incoming-Port", "wlan-debug")
+        ):
+            continue
+        if vids and (port.vid is None or port.vid not in vids):
+            continue
+        if pids and (port.pid is None or port.pid not in pids):
+            continue
+        if names and (port.name is None or not substr_in_list(port.name, names)):
```
Comment (dobairoland) [inline esptool/__init__.py:1062]: This substr_in_list is very specific and with inverting the result it is pretty hard to understand.  Could we replace it with an inline check similar to the following?

```python
... or any(name not in port.name for name in names)
```

File: esptool/__init__.py [hunk 3]
```diff
@@ -123,6 +123,14 @@ def main(argv=None, esp=None):
         default=os.environ.get("ESPTOOL_BAUD", ESPLoader.ESP_ROM_BAUD),
     )
 
+    parser.add_argument(
+        "--port-filter",
+        action="append",
+        help="Serial port device filter, can be vid=INT, pid=INT, name=SUBSTRING",
```
Comment (dobairoland) [inline esptool/__init__.py:129]: (optional) I think NUMBER instead of INT would be clearer for users.

**Review by dobairoland — CHANGES_REQUESTED** [review]
Thanks. This looks pretty good. I've left a couple of last suggestions.

**Comment by bryghtlabs-richard** [comment]
Thanks Roland, I've updated the commit with your suggestions.

File: esptool/__init__.py (hunk 2 above)
Comment (dobairoland) [reply esptool/__init__.py]: And I see I've made a mistake. You corrected it to `not any()`. That could be `all()` instead. You can ignore this if you won't make any more changes to this PR.

File: docs/en/esptool/advanced-options.rst [hunk 4]
```diff
@@ -117,3 +117,18 @@ An example of this is available in the :ref:`merge_bin <merge-bin>` command desc
...
+
+``--port-filter <FilterType>=<FilterValue>`` allows limiting ports that will be tried. This can be useful when esptool is run on a system
+with many serial ports. There are a few different types that can be combined. A port must match all specified FilterTypes, and must match
+at least one FilterValue for each specified FilterType to be considered. Example filter configurations:
+
+.. list::
+
+    * ``--port-filter vid=0x303A`` matches ports with the Espressif USB VID.
+    * ``--port-filter vid=0x303A --port-filter vid=0x0403`` matches Espressif and FTDI ports by VID.
+    * ``--port-filter vid=0x303A --port-filter pid=0x1001`` matches Espressif ESP32-S3 ports by VID and PID.
```
Comment (radimkarnis) [inline docs/en/esptool/advanced-options.rst:133]: `303A`:`1001` is not actually just the  ESP32-S3, but the USB-Serial/JTAG unit in general. That means any chip with this peripheral will show up with this VID:PID (S3, C3, C6, H2, P4, ...). 

If you want to mention a specific chip, USB-OTG mode has to be used. e.g. `ESP32-S2` = `VID:PID=303A:0002`

**Comment by radimkarnis** [comment]
Suggestion: A nice touch would be to add a mention about the Espressif USB customer-allocated PID repository
https://github.com/espressif/usb-pids. Essentially we allow users to register a PID under the Espressif VID for a specific piece of hardware. So one of the examples could maybe mention this?

Feel free to ignore, though!

**Comment by Jason2866** [comment]
Agree, nice feature!

**Comment by dobairoland** [comment]
@bryghtlabs-richard Thank you for the contribution. It was merged.

**Pull request merged**